enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+* Preflight checks should not mutate any existing resources as it will make the checks non-idempotent. So it is advisable to give read only access to preflight checks.
+
+
+#### Registering for preflight checks
+
+Operators need to register themselves for a preflight check by defining a Kubernetes job (v1) manifest.
+Preflight checks will be skipped for operator which does not have the manifest.
+
+An example of preflight job manifest
+```
+$ cat 0000_50_cloud-credential-operator_11_preflight-job.yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: cloud-credential-preflight-checks
+  annotations:
+    openshift.io/preflight-check: "true"
+spec:
+  template:
+    spec:
+      serviceAccountName: openshift-cloud-credential-preflight
+      containers:
+      - name: cloud-credential-operator
+        image: quay.io/openshift/origin-cloud-credential-operator:latest
+          resources:
+            requests:
+               cpu: 10m
+              memory: 150Mi
+        command: ["/bin/preflight",  "foo", "baar"]
+  backoffLimit: 4
+```
+
+
+*Note:*
+To give specific access to preflight jobs, you need to create a new ServiceAccount which the job can use and give the ServiceAccount required RBAC.
+
+#### Triggering execution of preflight checks
+
+We should be able to run the preflight check as below
+
+##### Case1
+
+Preflight checks would run implicitly right after the update is selected, as we currently do for preconditions checks.
+
+##### Case 2
+
+Run preflight checks on-demand for cluster admins using `--dry-run`.
+Example:
+
+```sh
+oc adm upgrade --to-latest --dry-run
+```
+
+#### Collecting results from preflight checks
+
+The preflight job for a release will be created with an unique label i.e. the set of preflight jobs triggered for doing a preflight check for a specific version will have the unique labels.
+This unique label will be used to keep track of the jobs and collect result from these.
+Also at any point of time there might be many preflight jobs running in the cluster.
+So it would also help to keep track of the preflight jobs based on when it was started.
+To collect and display the result we can query for all the jobs by the label.
+
+The objective is to collect the results i.e. pass or fail status and error message from preflight jobs.
+
+### Risks and Mitigation
+
+The responsibility of using implementation of actual preflight tests are on respective operator developers. So any bugs in the tests can create false positives or false negative results.
+
+## Design Details
+
+### Open Questions
+
+1. Should we save the preflight history? Will it be really useful for users?
+2. Tf the pre-flight is already running for a specific release and user again starts the preflight for the same image (same signature) we can just tell the user that the preflight is already running and here are the results so far, right?
```

Comment:
Not covered yet in this proposal is how old preflight results can be before we consider them stale (and reap the job?).  I'm fine just recycling the running job and/or old results as long as we haven't hit that threshold (24h?).

Reply:
I am thinking just reusing the job till it is in flight. Once a job finishes a new job will be created if user wants to do a prefligth check. Do not want to keep it complicated at this point of time.